### ISSUE TITLE
can't add community repo - "invalid location" error - problem solved

### ISSUE THREAD

**Issue by SlalomJohn** [SynoCommunity/spksrc#4897]
Synlogy RS815RP+ (2 node in cluster mode)

Adding repo as https://packages.synocommunity.com/
But got error
![image](https://user-images.githubusercontent.com/65710644/136002379-f5208622-5a48-4924-9070-6c85e763647d.png)

if i open this link in browser - got 
`Bad Request
The browser (or proxy) sent a request that this server could not understand.`

![image](https://user-images.githubusercontent.com/65710644/136002779-f2e547c8-d4a2-4410-9544-8ee78f0b7989.png)

**Comment by BenjV** [comment]
That link can only be used as package source and cannot be opened in a browser.
That said you most likely have a netwerk issue on your side so the package source cannot be reached by DSM.
My Nas does not have any problem to reached the SynCommunity packages.

**Comment by SlalomJohn** [comment]
> That link can only be used as package source and cannot be opened in a browser. That said you most likely have a netwerk issue on your side so the package source cannot be reached by DSM. My Nas does not have any problem to reached the SynCommunity packages.

> bash-4.3# ping packages.synocommunity.com
PING d.sni.global.fastly.net (151.101.246.133) 56(84) bytes of data.
64 bytes from 151.101.246.133 (151.101.246.133): icmp_seq=1 ttl=51 time=21.1 ms
64 bytes from 151.101.246.133 (151.101.246.133): icmp_seq=2 ttl=51 time=23.1 ms
64 bytes from 151.101.246.133 (151.101.246.133): icmp_seq=3 ttl=51 time=21.7 ms
64 bytes from 151.101.246.133 (151.101.246.133): icmp_seq=4 ttl=51 time=22.3 ms
64 bytes from 151.101.246.133 (151.101.246.133): icmp_seq=5 ttl=51 time=21.2 ms
^C
--- d.sni.global.fastly.net ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 4004ms
rtt min/avg/max/mdev = 21.126/21.915/23.156/0.752 ms
bash-4.3#

All internet sites opening. Ping work....
Previously worked at http://packages.synocommunity.com/
But at some point, it stopped downloading data from the community repository
It might be related to the old version DSM ?

**Comment by SlalomJohn** [comment]
> 2021-10-05T14:19:03+03:00 rs815rpplus synoscgi_SYNO.Core.Package.Feed_1_add[30612]: pkgcurltool.cpp:280 Failed to curl perform, code=60, err=Error
2021-10-05T14:19:03+03:00 rs815rpplus synoscgi_SYNO.Core.Package.Feed_1_add[30612]: pkgcurltool.cpp:453 Failed to download http://packages.synocommunity.com, code=60, err=Error

This in log's

**Comment by waspsoton** [comment]
I am having the same problem

**Comment by jedisamui** [comment]
I also get the "invalid location" error when trying to add the repo. Indeed, it *was* working before. I thought by removing/readding would resolve the issue, but alas, it does not. And now.... I cannot add back. 

From an SSH Session into my NAS (DS1816+), I performed the following: 

`$ curl -k https://packages.synocommunity.com
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>400 Bad Request</title>
  </head>
  <body>
    <h1>Bad Request</h1>
    <p>The browser (or proxy) sent a request that this server could not understand.</p>
  </body>
</html>`

`curl http://packages.synocommunity.com
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>400 Bad Request</title>
  </head>
  <body>
    <h1>Bad Request</h1>
    <p>The browser (or proxy) sent a request that this server could not understand.</p>
  </body>
</html>`

`sudo ping packages.synocommunity.com
Password:
PING d.sni.global.fastly.net (146.75.46.133) 56(84) bytes of data.
64 bytes from 146.75.46.133 (146.75.46.133): icmp_seq=1 ttl=55 time=10.2 ms
64 bytes from 146.75.46.133 (146.75.46.133): icmp_seq=2 ttl=55 time=10.0 ms
64 bytes from 146.75.46.133 (146.75.46.133): icmp_seq=3 ttl=55 time=9.60 ms
^C
--- d.sni.global.fastly.net ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2002ms
rtt min/avg/max/mdev = 9.602/9.986/10.274/0.305 ms`

`sudo nslookup packages.synocommunity.com
Server:		192.168.50.24
Address:	192.168.50.24#53

Non-authoritative answer:
packages.synocommunity.com	canonical name = d.sni.global.fastly.net.
Name:	d.sni.global.fastly.net
Address: 146.75.46.133`


It appears that I can not only reach you, but I can resolve your DNS name as well. I've read that this has happened in the past, and it was a change in your html code. Can you check you end and validate any changes made?

**Comment by Noone2018b** [comment]
Same issue here, but a couple of other observations suggest it is a network issue which I don't understand...

- It was previously working on my home, and a remote, DS1517+ unit.
- The remote unit is still OK, but my home unit gives the same issues as listed above. 
- On both units ping & nslookup are OK and I can access the website from a computer on the same network.
- On both units the curl test above (`curl http://packages.synocommunity.com`) gives a 400 result
- On the local unit I can't see any packages in the Community section of the package manager (i.e. in an already added package source), and I also can't connect as a new package source ("invalid location" error message).
- On the remote unit everything seems OK, I can see packages and also connect as a new package source.

UPDATED & FIXED: now fixed on my home unit after updating DSM to 6.2.4-25556 Update 2 (was DSM 6.2.3-something) and rebooting. (Although I'm not sure if it was a DSM version issue or just a clean boot would have fixed things!)

**Comment by markhofmeijer** [comment]
Same issue here (DSM 6.2.4-25556 Update 2). A reboot unfortunately does not fix it for me.

**Comment by jedisamui** [comment]
well. I be damned..... @Noone2018b - I updated my NAS, and that seems to have resolved it for me as well.

**Comment by SlalomJohn** [comment]
> finally I solved the issue by downloading the current certs.pem from openssl site (haxxe...) and saving this under the name ca-certificates.crt on the diskstation.

Are you send my direct link to **current certs.pem from openssl site** ? A can't find this cert....

**Comment by BenjV** [comment]
If it is a expired certificate problem you should renew it via the control panel, security, certificate and if it is expired right click to renew it.

**Comment by SlalomJohn** [comment]
> If it is a expired certificate problem you should renew it via the control panel, security, certificate and if it is expired right click to renew it.

Not work. This operation created request, but not renew or update current cert's

**Comment by BenjV** [comment]
https://kb.synology.com/en-af/DSM/help/DSM/AdminCenter/connection_certificate?version=6

**Comment by hgy59** [comment]
> My file is:
> ca-certificates.crt │ 274340│May 10 2019

my one was from 2016. So I'm not shure it is the same problem. Do you get the error 60?

**Comment by hgy59** [comment]
> Are you send my direct link to **current certs.pem from openssl site** ? A can't find this cert....

try this one https://curl.se/ca/cacert.pem

**Comment by hgy59** [comment]
>  you can update the trust store manually by overriding the file with a more recent one (assuming you trust the curl developers),

Yes, everybody should update the DSM. My solution is for DS-210+ and other old models that do not support DSM 6.

The cacert.pem you download from curl.se is not managed by curl, but it is the pem version of the mozillas trust store, downloaded and converted by mk-ca-bundle (see https://curl.se/docs/mk-ca-bundle.html). So everything is transparent as you can inspect and execute mk-ca-bundle yourself.
Indeed our community packages included this bundle with all packages that contain curl, before we switched to the DSM integrated version to avoid the need to update packages just to update the ca-bundle (beliving that synology does the update).

**Comment by publicarray** [comment]
@hgy59 I've send an email to Synology asking how long their release rollout is. People still run into this issue about a month after the initial release of their update. I mean how long do we keep this issue open?

**Comment by mfriedrichs** [comment]
IF it is a certifcate-related error, this MAY help:

I copied this file (that was mentioned before by  @hgy59 ):
https://curl.se/ca/cacert.pem
to these locations:
/usr/syno/etc/certificate/system/default/
/usr/syno/etc/certificate/system/FQDN/
and renamed them to "cert.pem", after i moved the old "cert.pem" files to a backup-folder (just in case something went wrong).
After that, i navigated INTO these folders and gave the "cert.pem" files root permissions:
chown root:root cert.pem

At first, nothing changed, even after e reboot. 
But when i rebooted the DSM for different reasons a few times and after a while, the repository-list was loaded.  

I didn't know where to copy this file until i found this thread, where the author  ( @catchdave ) also offers a script for automation:
https://gist.github.com/catchdave/69854624a21ac75194706ec20ca61327#file-replace_synology_ssl_certs-sh 

Hope, this will help some of you.

**Comment by publicarray** [comment]
Please don't do this. You just replaced the Synology's self signed certificate with the combined trust store from Mozilla.

If you for what ever reason can't update the DSM please read the comment above: https://github.com/SynoCommunity/spksrc/issues/4897#issuecomment-937376994

If it works again it's probably because the NAS updated it self.

**Comment by mfriedrichs** [comment]
> Please don't do this. 
[...]
> If you for what ever reason can't update the DSM please read the comment above: [#4897 (comment)](https://github.com/SynoCommunity/spksrc/issues/4897#issuecomment-937376994)

So,- would you also be so kind to explain, WHY this should be the way, to integrate the cacert.pem from curl.se and not otherwise?
(btw: the discord link leads into emptiness and - sorry- but the other link was way too technical and too far from my understanding and seems barely related to this question).

> If it works again it's probably because the NAS updated it self.

I really don't think so. If so,- it would be solved for everyone just through waiting by now.
I would be curious to know your answer to my question.

**Comment by publicarray** [comment]
Sure you use any method to update your cert. But `cacert.pem` is not a certificate. It's the root trust store, a list of trusted Certificate Authorities (CA) for your operating system. (it's in the name CAcert.pem). Synology was a bit slow to update theirs and their rollout is really slow. File location for the trust store is in `/etc/ssl/certs/ca-certificates.crt` as mentioned above: https://github.com/SynoCommunity/spksrc/issues/4897#issuecomment-934744470 not `/usr/syno/etc/certificate/system/default/`

> I really don't think so. If so,- it would be solved for everyone just through waiting by now.

Yes the fix is to just update to **6.2.4-25556 Update 2** or later.
From the Change log: `10 Updated the OpenSSL setting in response to new rules of Let's Encrypt certificates.`

TLDR: The root course is that the Synology's outdated trust store no longer trusts Let's encrypt certificates because the `DST Root CA X3` expired. The fix is to update the trust store. 
https://github.com/SynoCommunity/spksrc/issues/4897#issuecomment-943705896

If you can't follow along I recommend reading more about PKI (Public Key Infrastructure), CA's (Certificate Authority), [Chain Of Trust](https://en.wikipedia.org/wiki/Chain_of_trust) and Public-key cryptography (Diffie–Hellman key exchange)

https://www.youtube.com/watch?v=fuK-OAyfET4

**Comment by hgy59** [comment]
>  Synology was a bit slow to update theirs and their rollout is really slow

Yes, they are sloooooooooow. I got the notification for the DSM 7.0.1 update yesterday (11/02/2021) - the update was released on 10/21/2021.

**Comment by acesabe** [comment]
Just ran into this issue (and been dealing with the fallout from the same CA expiry at work for weeks!) and confirm that upgrading to the latest version of DSM 6.2.4-25556 **Update 2** has indeed resolved the issue.
I did remove the SynoCommunity package repo beforehand tho in attempt to get the cert error fixed so worth trying that also if issue persists.

**Comment by SlalomJohn** [comment]
https://github.com/SynoCommunity/spksrc/issues/4897#issuecomment-937376994
`sudo mv /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt.bak && sudo curl -Lko /etc/ssl/certs/ca-certificates.crt https://curl.se/ca/cacert.pem`
This issue is fix error. Very thanks!

**Comment by publicarray** [comment]
> > Synology was a bit slow to update theirs and their rollout is really slow
> 
> Yes, they are sloooooooooow. I got the notification for the DSM 7.0.1 update yesterday (11/02/2021) - the update was released on 10/21/2021.

From communicating with Synology:

>  It usually takes 6~8 weeks to complete a rolling release

**Comment by mathmathou** [comment]
Hello,
I confirm that upgrading to version DSM 6.2.4-25556 Update 2 solved the issue for me as well.
After a reboot, I had no more error message when adding the package source.
DS 412+

**Comment by SlalomJohn** [comment]
> Hello, I confirm that upgrading to version DSM 6.2.4-25556 Update 2 solved the issue for me as well. After a reboot, I had no more error message when adding the package source. DS 412+

My two DSM is accepted certificates without reboot :)

**Comment by publicarray** [comment]
Please don't close the issue yet so that others can find this thread more easily. If you get too many notifications you can unsubscribe from the right sidebar.

**Comment by publicarray** [comment]
From Synology:

> Update2 has completed its rollout process, and everyone should've received that version.

**Comment by robigombek** [comment]
> > So yes after some testing on my end I can confirm that the trust certificates on a not fully updated DSM 6 are too old (a certificate must have expired recently).  If you want to continue to stay on an old versions (not recommended obviously) you can update the trust store manually by overriding the file with a more recent one (assuming you trust the curl developers), using SSH: `sudo mv /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt.bak && sudo curl -Lko /etc/ssl/certs/ca-certificates.crt https://curl.se/ca/cacert.pem` This will fix the issue, alternatively you can set the clock back. The best solution however is to update to a more recent DSM6 version.
> 
> https://discord.com/channels/732558169863225384/732559466721181738/893810613384515584
> 
> Might also be related to https://scotthelme.co.uk/lets-encrypt-old-root-expiration/

These commands do the job! Thank you!!!

**Comment by gingerbeardman** [comment]
> using SSH: `sudo mv /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt.bak && sudo curl -Lko /etc/ssl/certs/ca-certificates.crt https://curl.se/ca/cacert.pem` This will fix the issue, 

Thanks for the answer!

**Comment by Doublefire-Chen** [comment]
> > So yes after some testing on my end I can confirm that the trust certificates on a not fully updated DSM 6 are too old (a certificate must have expired recently).  If you want to continue to stay on an old versions (not recommended obviously) you can update the trust store manually by overriding the file with a more recent one (assuming you trust the curl developers), using SSH: `sudo mv /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt.bak && sudo curl -Lko /etc/ssl/certs/ca-certificates.crt https://curl.se/ca/cacert.pem` This will fix the issue, alternatively you can set the clock back. The best solution however is to update to a more recent DSM6 version.
> 
> https://discord.com/channels/732558169863225384/732559466721181738/893810613384515584
> 
> Might also be related to https://scotthelme.co.uk/lets-encrypt-old-root-expiration/

Thank you! It works on my DSM 6.2.1

**Comment by florisgravendeel** [comment]
I've had the same issue on DSM 6.2.1. 

I performed a manual software update to renew the certificates. Updating to DSM 7.1.1 solved the problem for me.

**Comment by debuglevel** [comment]
For the shy which do not want to update to DSM7 yet: Updated from DSM 6.3.x to 6.4.y and works now.

**Comment by gingerbeardman** [comment]
@debuglevel please tell me more
